// @bundle-stats/utils/lib-esm/webpack/extract/modules.js
import isEmpty from 'lodash/isEmpty';
import round from 'lodash/round';
import { getModuleName, normalizeChunkId } from '../utils';
export const extractModules = (webpackStats) => {
    const modulesSource = (webpackStats === null || webpackStats === void 0 ? void 0 : webpackStats.modules) || [];
    if (!modulesSource) {
        return {
            metrics: {
                duplicateCode: { value: 0 },
                duplicateModulesCount: { value: 0 },
                modules: {},
            },
        };
    }
    const allModules = modulesSource.reduce((agg, moduleEntry) => {
        if (!moduleEntry.modules) {
            agg.push(moduleEntry);
            return agg;
        }
        agg = agg.concat(moduleEntry.modules.map((concatenatedModule) => ({
            ...concatenatedModule,
            chunks: moduleEntry.chunks,
        })));
        return agg;
    }, []);
    let totalCodeSize = 0;
    let duplicateCodeSize = 0;
    let duplicateModulesCount = 0;
    const modules = allModules.reduce((agg, moduleEntry) => {
        const { name, size, chunks } = moduleEntry;
        const normalizedName = getModuleName(name);
        if (isEmpty(chunks)) {
            return agg;
        }
        const instances = chunks.length;
        const duplicateInstances = instances - 1;
        duplicateModulesCount += duplicateInstances;
        duplicateCodeSize += duplicateInstances * size;
        totalCodeSize += instances * size;
        agg[normalizedName] = {
            name,
            value: size,
            chunkIds: chunks.map(normalizeChunkId),
            duplicated: Boolean(duplicateInstances),
        };
        return agg;
    }, {});
    const duplicateCode = totalCodeSize ? round((duplicateCodeSize / totalCodeSize) * 100, 2) : 0;
    return {
        metrics: {
            modules,
            duplicateCode: {
                value: duplicateCode,
            },
            duplicateModulesCount: {
                value: duplicateModulesCount,
            },
        },
    };
};
